Source/bmalloc/libpas/src/libpas/pas_stats.h

```diff
@@ -0,0 +1,311 @@
+// to log which significantly distorts the stat logging functionality.
+#define PAS_STATS_MALLOC_INFO_HIGH_GRANULARITY_MAX_SIZE   (1ull << 12)
+#define PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY_MAX_SIZE (1ull << 15)
+#define PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_MAX_SIZE    (1ull << PAS_VA_BASED_ZERO_MEMORY_SHIFT)
+
+#define PAS_STATS_MALLOC_INFO_HIGH_GRANULARITY_SHIFT      PAS_MIN_ALIGN_SHIFT
+#define PAS_STATS_MALLOC_INFO_HIGH_GRANULARITY            (1ull << PAS_STATS_MALLOC_INFO_HIGH_GRANULARITY_SHIFT)
+#define PAS_STATS_MALLOC_INFO_HIGH_GRANULARITY_BUCKETS    (PAS_STATS_MALLOC_INFO_HIGH_GRANULARITY_MAX_SIZE / PAS_STATS_MALLOC_INFO_HIGH_GRANULARITY)
+#define PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY_SHIFT    PAS_MIN_MEDIUM_ALIGN_SHIFT
+#define PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY          (1ull << PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY_SHIFT)
+#define PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY_BUCKETS  (PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY_MAX_SIZE / PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY)
+#define PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_SHIFT       PAS_MIN_MARGE_ALIGN_SHIFT
+#define PAS_STATS_MALLOC_INFO_LOW_GRANULARITY             (1ull << PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_SHIFT)
+#define PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_BUCKETS     (PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_MAX_SIZE / PAS_STATS_MALLOC_INFO_LOW_GRANULARITY)
+// FIXME: handle larger sizes in a sane way -- powers of two?
+#define PAS_STATS_MALLOC_INFO_OVERSIZE_BUCKETS            (1ull)
+
+#define PAS_STATS_MALLOC_INFO_BUCKET_COUNT_PER_STAT (\
+    PAS_STATS_MALLOC_INFO_HIGH_GRANULARITY_BUCKETS + \
+    PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY_BUCKETS + \
+    PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_BUCKETS + \
+    PAS_STATS_MALLOC_INFO_OVERSIZE_BUCKETS)
+
+typedef struct {
+    pas_stats_stat_base base;
+    uint64_t total_count;
+    uint64_t count_by_heap_type[pas_stats_heap_type_count];
+    uint64_t count_by_size[PAS_STATS_MALLOC_INFO_BUCKET_COUNT_PER_STAT];
+} pas_stats_malloc_info_data;
+
+#if PAS_ENABLE_STATS
+
+char* pas_stats_malloc_info_log(void* stat_v);
+void pas_stats_malloc_info_record(pas_stats_malloc_info_data* data, pas_stats_heap_type heap_type, size_t size, size_t count);
+
+#define PAS_STATS_FOR_EACH_COUNTER(OP) \
+    OP(malloc_info_bytes, pas_stats_malloc_info_data, pas_stats_malloc_info_log) \
+    OP(malloc_info_allocations, pas_stats_malloc_info_data, pas_stats_malloc_info_log)
+
+// FIXME: in principle it should be possible to automatically generate this via PAS_STATS_FOR_EACH_COUNTER, somehow
+#define PAS_RECORD_STAT_malloc_info_bytes(data, heap_type, size) \
+    pas_stats_malloc_info_record(data, heap_type, size, size)
+#define PAS_RECORD_STAT_malloc_info_allocations(data, heap_type, size) \
+    pas_stats_malloc_info_record(data, heap_type, size, 1)
+
+#else /* !PAS_ENABLE_STATS */
+#define PAS_RECORD_STAT(name, ...) do { PAS_UNUSED_V(__VA_ARGS__); } while (0)
```

Comment:
Please move this to the same `#if PAS_ENABLE_STATS` block below where `PAS_RECORD_STAT` is defined.  It is very disorienting and hard to read that the disabled version is here, while the enabled version is elsewhere.  Let's keep them paired together.